### DISCUSSION TITLE
Use rest parameters to make `expect.any()` variadic

### DISCUSSION THREAD

**Discussion by zirkelc** [vitest-dev/vitest#6961] [Ideas]
With `expect.any()` we can only have one type like `expect.any(String)`. If we want to express an optional property that might be `null` or `undefined`, we have to use a 3rd party library.

If `expect.any()` were variadic, it could be called with multiple possible types expressing a logical-OR:

```ts
const user = isNew ? { name: "John" } : { id: 1, name: "Jane" };
expect(user).toEqual({ id: expect.any(Number, undefined) }); // ID can be undefined or Number
```

Making the signature variadic shouldn't break anything, but it would be a slight derivation from the Jest-style API.

**Comment by sheremet-va** [comment]
I am not convinced that we should allow additional arguments, but maybe we can add a new function like `expect.oneOf(...args)`, so you example would looks like this:

```ts
expect(user).toEqual({
  id: expect.oneOf(expect.any(String), undefined))
})
```

**Reply by zirkelc** [reply]
Yes, that would be great! I wasn't sure you want to deviate from the Jest API, so I thought making `expect.any()` variadic is a minimally invasive solution.

**Reply by hi-ogawa** [reply]
It doesn't seem so from the doc https://jestjs.io/docs/expect#expectanyconstructor but does Jest support this? If so, then aligning with them should be fine.

Speaking of `oneOf`, chai seems to have something similar as non-asymmetric assertion https://www.chaijs.com/api/bdd/#method_oneof.

**Reply by zirkelc** [reply]
No, as far as I know Jest does _not_ support `expect.any(...args)`, but making this change wouldn't add another function (i.e. minimally invasive). 

The alternative `expect.oneOf(expect.any(String), undefined))` would be better IMO, but it would add a new function. If wasn't sure how strict you want to match the Jest APIs.

Let me know if you'd like me to try my hand at PR for this project.

**Reply by hi-ogawa** [reply]
Ah, got it :+1: You meant adding new `expect.oneOf` can be bigger deviation as it doesn't exists in Jest. For us, adding a new API is totally fine and encouraged when the new feature makes sense.

**Reply by zirkelc** [reply]
I created PR https://github.com/vitest-dev/vitest/pull/6974 for this, let me know what you think! I used an array paramater for `expect.oneOf([ ... ])` instead of variadic args, but I'm open to changing it.
